pkg/common/commands/resetoffset/controller/dataplane.go

```diff
@@ -0,0 +1,215 @@
+	podIP string,
+	service ctrl.Service,
+	refInfo *refmappers.RefInfo,
+	opCode ctrl.OpCode) error {
+
+	// Get The Logger From Context
+	logger := logging.FromContext(ctx).Desugar().With(zap.String("PodIP", podIP), zap.Int("OpCode", int(opCode)))
+
+	// Generate A CommandID For The ResetOffset
+	commandId, err := generateCommandId(resetOffset, podIP, opCode)
+	if err != nil {
+		logger.Error("Failed to generate Command ID for ResetOffset", zap.Error(err))
+		return fmt.Errorf("failed to generate Command ID for ResetOffset: %v", err)
+	}
+
+	// Create And Send A The ConsumerGroupAsyncCommand to The Control-Protocol Service
+	consumerGroupAsyncCommand := commands.NewConsumerGroupAsyncCommand(commandId, refInfo.TopicName, refInfo.GroupId)
+	err = service.SendAndWaitForAck(opCode, consumerGroupAsyncCommand)
+	if err != nil {
+		logger.Error("Failed to send ConsumerGroup AsyncCommand", zap.Int64("CommandID", commandId), zap.Error(err))
+		return fmt.Errorf("failed to send ConsumerGroup AsyncCommand '%d' : %v", commandId, err)
+	}
+
+	// Wait For The AsyncCommand Result & Return Results
+	return r.waitForAsyncCommandResult(resetOffset, podIP, consumerGroupAsyncCommand)
+}
+
+// waitForAsyncCommandResult polls the Reconciler AsyncCommandNotificationStore waiting for the
+// AsyncCommandResult corresponding to the specified AsyncCommand.  The ConsumerGroupAsyncCommands
+// are inherently asynchronous so that they can be used in other scenarios (Pause/Resume), but the
+// ResetOffset implementation treats them as Synchronous to facilitate single-pass reconciliation.
+func (r *Reconciler) waitForAsyncCommandResult(resetOffset *kafkav1alpha1.ResetOffset, podIP string, asyncCommand ctrlmessage.AsyncCommand) error {
+
+	// Create A NamespacedName For The ResetOffset
+	resetOffsetNamespacedName := types.NamespacedName{
+		Namespace: resetOffset.GetNamespace(),
+		Name:      resetOffset.GetName(),
+	}
+
+	// Poll The AsyncCommandNotificationStore For AsyncCommandResult
+	err := wait.Poll(asyncCommandResultPollDuration, asyncCommandResultTimeoutDuration, func() (done bool, err error) {
+		asyncCommandResult := r.asyncCommandNotificationStore.GetCommandResult(resetOffsetNamespacedName, podIP, asyncCommand)
+		if asyncCommandResult != nil {
+			if asyncCommandResult.IsFailed() {
+				return true, fmt.Errorf("AsyncCommand ID '%x' resulted in error: %s", asyncCommand.SerializedId(), asyncCommandResult.Error)
+			} else {
+				return true, nil // Return Success
+			}
+		} else {
+			return false, nil // Not Found - Try Again
+		}
+	})
+
+	// Return The Result
+	return err
+}
+
+// generateCommandId returns an int64 hash based on the specified ResetOffset.
+func generateCommandId(resetOffset *kafkav1alpha1.ResetOffset, podIP string, opCode ctrl.OpCode) (int64, error) {
+	hash := fnv.New32a()
+	_, err := hash.Write([]byte(fmt.Sprintf("%s-%d-%s-%d", string(resetOffset.UID), resetOffset.Generation, podIP, opCode)))
+	if err != nil {
```

Comment:
The sum32 implementation of the Hash32 Write() call can't actually return an error (though it's probably not bad to check it anyway; just mentioning it - might simplify the generateCommandId call chain).

Reply:
Yeah, seems like bad form to code against the inner implementation of a third party library rather than against the interface.  Good future foot-gun if the implementation changes in the future and starts returning errors ; )